### DISCUSSION TITLE
Count active subscriptions

### DISCUSSION THREAD

**Discussion by AlexanderWells-diamond** [strawberry-graphql/strawberry#2719] [Q&A] [answered]
I'm trying to get a count of the number of active subscriptions in my Strawberry server. I have been unable to find how to do this. For reference I'm using an `aiohttp` server.

I've tried a variety of methods:
- Using custom extensions. These do not seem to trigger for subscriptions at all. 
- Examining the various Tracing extensions. None of these appear to offer subscription tracking.
- Looking through Strawberry code. I found the `BaseGraphQLTransportWSHandler` class, which seems to have a [dictionary of subscriptions](https://github.com/strawberry-graphql/strawberry/blob/3310d6c87bdea46ab1ac6ca16fdea17dd011b5e6/strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py#L50). However, it doesn't seem to be possible to access this - the `GraphQLView` that we pass to `aiohttp` [dynamically creates new instances for every request](https://github.com/strawberry-graphql/strawberry/blob/3310d6c87bdea46ab1ac6ca16fdea17dd011b5e6/strawberry/aiohttp/views.py#L136) , meaning that there's no way to access this data from the `GraphQLView` (or anywhere for that matter as the reference is lost)
- Looking at `aiohttp` itself, but it does not seem to support any websocket connection tracing.

Is there something I've missed? Thanks for any help!

**Comment by DoctorJohn** [comment, marked as answer]
Hey @AlexanderWells-diamond, currently I can't think of an trivial way to achieve this. The aiohttp view creates a new instance of `strawberry.aiohttp.handles.GraphQLTransportWSHandler` for each incoming ws connection using the graphql-transport-ws protocol. You could subclass said class and extend its `handle_request` method to keep track of individual connections by updating some global counter.

The code could look roughtly like this:

```python
from strawberry.aiohttp.handlers import GraphQLTransportWSHandler
from strawberry.aiohttp.views import GraphQLView
from aiohttp import web

some_global_counter = 0


class CustomHandler(GraphQLTransportWSHandler):
  async def handle_request(self):
    some_global_counter = some_global_counter + 1
    await super().handle_request()
    some_global_counter = some_global_counter - 1


class CustomView(GraphQLView):
  graphql_transport_ws_handler_class = CustomHandler


schema = ... # your schema

app = web.Application()
app.router.add_route("*", "/graphql", GraphQLView(schema=schema))
```

Note that I haven't tested it, just wanted to give you a starting point to a potential solution.

**Reply by AlexanderWells-diamond** [reply]
Thank you for the suggestion! I'll give it a try

**Reply by AlexanderWells-diamond** [reply]
Thank you for this, it seems to do exactly what I wanted.

For anyone else who stumbles across this: I was trying to set up metrics tracking with [aioprometheus](https://aioprometheus.readthedocs.io/en/latest/index.html), specifically using its `@inprogress` decorator. This code seems to work as you'd expect to track subscriptions:

```
class MetricsGraphQLTransportWSHandler(GraphQLTransportWSHandler):
    @inprogress(
        REQUESTS_IN_PROGRESS, labels={"type": "subscription_GraphQLTransportWS"}
    )
    async def handle_request(self):
        await super().handle_request()
```
